Makes the explosive compressor and blastcannon actually use the TTVs they're given + the explosion changes to support that.

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Adds a simple framework to let objects respond to explosions occurring inside of them. 
- Changes a whole bunch of explosions to use the object being exploded as the origin of the explosion rather than the turf the object is on.
- Makes the explosive compressor and blastcannon actually use the TTVs they are given.

- [X] TODO: Less ugly argument passing

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- Adds support for things responding to internal explosions.
- Less snowflake code for the explosive compressor and blastcannon calculating bomb range.*
- Less confusing explosion severity defines.
- Less opaque explosion arguments
- Fixes #58717 

*does not guarantee that the solution to letting them actually use the TTV is any less snowflake.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
refactor: The explosive compressor and blastcannon now actually use the bombs they are given. Both have experienced slight changes to their response times as a result.
code: The explosion severity definitions are now the right way round.
code: Most explosions now use named arguments and occur at the exploding object instead of its turf.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Initial Commit

TemporalOroboros committed:
Adds explosion SFX to the blastcannon and explosive compressor

- Extracts the explosion SFX and screenshake proc from the SSexplosions explosion handling proc and lets the explosive compressor and blastcannon use it.

ATH1909 commented:
does this happen to fix any other explosion bugs, like some of the fireball ones?

TemporalOroboros commented:
@ATH1909 
> 
> 
> does this happen to fix any other explosion bugs, like some of the fireball ones?

I don't know of the explosion bugs you're referring to so probably not.

moocowswag commented:
Id consider this a balance change to blast cannons since you cant reuse tank transfer valves after this change.

TemporalOroboros commented:
@moocowswag 
> Id consider this a balance change to blast cannons since you wont be able to reuse tank transfer valves and spam short range shots.

Actually no, since I didn't fix this. They will eat one of the tanks and empty the other one though.

> But then again they probably wernt intended to be reusable.

Yes. But I don't care that they weren't. It's funny.

> And i think one of the bugs athath is refering to is how fireballs wont actually explode when they hit a bot like beepsky and instead will somehow only damage the bot as if it was just a damaging projectile.

I'll check it once I fix my copy of DM refusing to debug anything for some reason.

TemporalOroboros committed:
Merge branch 'master' into internal-explosions

TemporalOroboros committed:
Miscellaneous changes

- Adds defines for the internal explosion arglist keys
- Reverses the values of the explosion severity defines
- Changes almost everything that uses `/proc/explosion` to use named arguments

TemporalOroboros committed:
Actually invert these

TemporalOroboros commented:
@ATH1909 @moocowswag 
I think I figured out the fireball bug. The fireball does its damage before it tries to handle the explosion. Bots delete themselves when they die. As a result, if the fireball kills the bot before it gets around to the exploding bit it winds up trying to explode null and it breaks.

LemonInTheDark commented:
aw shit here we go again

TemporalOroboros committed:
Merge branch 'master' into internal-explosions

LemonInTheDark commented on code/game/gamemodes/meteor/meteors.dm:
```diff
@@ -276,7 +276,7 @@ GLOBAL_LIST_INIT(meteorsC, list(/obj/effect/meteor/dust)) //for space dust event
 /obj/effect/meteor/irradiated/meteor_effect()
 	..()
-	explosion(src.loc, 0, 0, 4, 3, 0)
+	explosion(src, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 4, flame_range = 0, flash_range = 3, adminlog = FALSE)
```
ditto

TemporalOroboros replied:
Done

LemonInTheDark commented on code/game/gamemodes/meteor/meteors.dm:
```diff
@@ -350,12 +350,12 @@ GLOBAL_LIST_INIT(meteorsC, list(/obj/effect/meteor/dust)) //for space dust event
 /obj/effect/meteor/tunguska/meteor_effect()
 	..()
-	explosion(src.loc, 5, 10, 15, 20, 0)
+	explosion(src, devastation_range = 5, heavy_impact_range = 10, light_impact_range = 15, flame_range = 0, flash_range = 20, adminlog = FALSE)
```
and again

TemporalOroboros replied:
Done

LemonInTheDark commented on code/game/gamemodes/meteor/meteors.dm:
```diff
@@ -350,12 +350,12 @@ GLOBAL_LIST_INIT(meteorsC, list(/obj/effect/meteor/dust)) //for space dust event
 	..()
 	if(prob(20))
-		explosion(src.loc,2,4,6,8)
+		explosion(src, devastation_range = 2, heavy_impact_range = 4, light_impact_range = 6, flame_range = 0, flash_range = 8, adminlog = FALSE)
```
ONE MORE TIME

TemporalOroboros replied:
Done

LemonInTheDark commented on code/modules/events/spacevine.dm:
```diff
@@ -115,7 +115,7 @@
 
 /datum/spacevine_mutation/explosive/on_death(obj/structure/spacevine/holder, mob/hitter, obj/item/I)
-	explosion(holder.loc, 0, 0, severity, 0, 0)
+	explosion(holder, devastation_range = 0, heavy_impact_range = 0, light_impact_range = severity, flash_range = 0, adminlog = FALSE)
```
why set flash range?

TemporalOroboros replied:
Consistency I guess.

TemporalOroboros replied:
Removed

LemonInTheDark reviewed (changes requested):
Here's just the conversions, I'll do the 4 major file changes later

TemporalOroboros commented on code/modules/vehicles/mecha/working/ripley.dm:
```diff
@@ -180,7 +180,7 @@
 /obj/vehicle/sealed/mecha/working/ripley/contents_explosion(severity, target)
 	for(var/i in cargo)
 		var/obj/cargoobj = i
-		if(prob(30/severity))
+		if(prob(10 * severity))
```
The defines were reversed. I un-reversed them and had to compensate.

TemporalOroboros commented on code/modules/power/rtg.dm:
```diff
@@ -76,7 +76,7 @@
 	playsound(src.loc, 'sound/magic/lightningshock.ogg', 100, TRUE, extrarange = 5)
 	tesla_zap(src, 5, power_gen * 0.05)
-	addtimer(CALLBACK(GLOBAL_PROC, .proc/explosion, get_turf(src), 2, 3, 4, 8), 100) // Not a normal explosion.
+	addtimer(CALLBACK(GLOBAL_PROC, .proc/explosion, src, 2, 3, 4, null, 8), 100) // Not a normal explosion.
```
The null was from before I started named args-ing everything. I'll fix it.

TemporalOroboros replied:
Apparently you can't use named arguments in callbacks. The null stays.

TemporalOroboros commented on code/modules/power/gravitygenerator.dm:
```diff
@@ -38,7 +38,7 @@ GLOBAL_LIST_EMPTY(gravity_generators) // We will keep track of this by adding ne
 
 /obj/machinery/gravity_generator/ex_act(severity, target)
-	if(severity == EXPLODE_DEVASTATE) // Very sturdy.
+	if(severity >= EXPLODE_DEVASTATE) // Very sturdy.
 		set_broken()
```
There is nothing above `EXPLODE_DEVASTATE`. If someone adds something stronger at some point it should still break the generator.

TemporalOroboros commented on code/modules/mob/living/simple_animal/hostile/megafauna/demonic_frost_miner.dm:
```diff
@@ -151,7 +151,7 @@ Difficulty: Extremely Hard
 
 /mob/living/simple_animal/hostile/megafauna/demonic_frost_miner/ex_act(severity, target)
-	adjustBruteLoss((30 * severity) - 120)
+	adjustBruteLoss(-30 * severity)
```
Yep.

TemporalOroboros commented on code/modules/mining/machine_vending.dm:
```diff
@@ -200,7 +200,7 @@
 
 /obj/machinery/mineral/equipment_vendor/ex_act(severity, target)
 	do_sparks(5, TRUE, src)
-	if(prob(50 / severity) && severity < EXPLODE_LIGHT)
+	if(severity > EXPLODE_LIGHT && prob(17 * severity))
 		qdel(src)
```
Yep

TemporalOroboros commented on code/modules/events/crystal_event.dm:
```diff
@@ -121,7 +121,7 @@ This section is for the event controller
 		center_areas += center_area
 
-		explosion(center_turf,0,0,5,7,7)
+		explosion(center_turf, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 5, flame_range = 7, flash_range = 7)
```
Nope. For some reason it was setting `adminlog = 7`. Since that's supposed to be a bool I assumed it was intended to be flame range.

LemonInTheDark replied:
![image](https://user-images.githubusercontent.com/58055496/115363608-76213500-a177-11eb-8480-8a532e9888bc.png)

TemporalOroboros commented on code/datums/diseases/rhumba_beat.dm:
```diff
@@ -39,4 +39,4 @@
 			to_chat(affected_mob, "<span class='danger'>Your body is unable to contain the Rhumba Beat...</span>")
 			if(DT_PROB(29, delta_time))
-				explosion(get_turf(affected_mob), -1, 0, 2, 3, 0, 2) // This is equivalent to a lvl 1 fireball
+				explosion(affected_mob, devastation_range = -1, heavy_impact_range = 0, light_impact_range = 2, flame_range = 2, flash_range = 3, adminlog = FALSE) // This is equivalent to a lvl 1 fireball
```
That's what I'm assuming. I moved it to flame range since that's the next thing in line that isn't a bool.

TemporalOroboros commented on code/datums/mutations/body.dm:
```diff
@@ -451,7 +451,7 @@
 		qdel(I)
 
-	explosion(get_turf(owner), 0, 0, 2, 0, TRUE)
+	explosion(owner, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 2, flame_range = 0, flash_range = 0, adminlog = TRUE)
```
I'll fix it.

TemporalOroboros replied:
Fixed

TemporalOroboros commented on code/modules/antagonists/changeling/powers/headcrab.dm:
```diff
@@ -18,7 +18,7 @@
 		I.Remove(user, 1)
 
-	explosion(get_turf(user), 0, 0, 2, 0, TRUE)
+	explosion(user, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 2, flash_range = 0, adminlog = TRUE)
```
Ditto

TemporalOroboros replied:
Done

TemporalOroboros commented on code/modules/admin/smites/bsa.dm:
```diff
@@ -10,7 +10,7 @@
 	. = ..()
 
-	explosion(target.loc, 0, 0, 0, 0)
+	explosion(target.loc, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 0, flash_range = 0)
```
I'm pretty sure this means it hits just that specific turf.

LemonInTheDark replied:
hhhhhhhhhhhhhhhhh

TemporalOroboros commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -126,7 +126,7 @@
 
 /obj/item/tank/deconstruct(disassembled = TRUE)
-	var/turf/location = get_turf(src)
+	var/turf/location = loc
```
Right, I need to change that cast.

TemporalOroboros replied:
Done

TemporalOroboros commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -251,7 +251,7 @@
 	if(QDELETED(src) || !leaking || !air_contents)
 		return
-	var/turf/location = get_turf(src)
+	var/turf/location = loc
```
Same

TemporalOroboros replied:
Done

TemporalOroboros commented on code/game/objects/items/granters.dm:
```diff
@@ -159,7 +159,7 @@
 /obj/item/book/granter/spell/fireball/recoil(mob/user)
 	..()
-	explosion(user.loc, 1, 0, 2, 3, FALSE, FALSE, 2)
+	explosion(user, devastation_range = 1, heavy_impact_range = 0, light_impact_range = 2, flame_range = 2, flash_range = 3, adminlog = FALSE)
```
Yeah, it was the default for `ignorecap`

TemporalOroboros committed:
Merge branch 'master' into internal-explosions

TemporalOroboros commented on code/controllers/subsystem/explosions.dm:
```diff
@@ -169,20 +169,85 @@ SUBSYSTEM_DEF(explosions)
+/datum/controller/subsystem/explosions/proc/explode(atom/origin, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 0, flame_range = 0, flash_range = 0, adminlog = TRUE, ignorecap = FALSE, silent = FALSE, smoke = FALSE)
+	var/list/arguments = list(EXARG_KEY_ORIGIN = origin, EXARG_KEY_DEV_RANGE = devastation_range, EXARG_KEY_HEAVY_RANGE = heavy_impact_range, EXARG_KEY_LIGHT_RANGE = light_impact_range, EXARG_KEY_FLAME_RANGE = flame_range, EXARG_KEY_FLASH_RANGE = flash_range, EXARG_KEY_ADMIN_LOG = adminlog, EXARG_KEY_IGNORE_CAP = ignorecap, EXARG_KEY_SILENT = silent, EXARG_KEY_SMOKE = smoke)
+	var/atom/location = isturf(origin) ? origin : origin.loc
+	if(SEND_SIGNAL(origin, COMSIG_ATOM_EXPLODE, arguments[EXARG_KEY_DEV_RANGE], arguments[EXARG_KEY_HEAVY_RANGE], arguments[EXARG_KEY_LIGHT_RANGE], arguments[EXARG_KEY_FLAME_RANGE], arguments[EXARG_KEY_FLASH_RANGE], arguments[EXARG_KEY_ADMIN_LOG], arguments[EXARG_KEY_IGNORE_CAP], arguments[EXARG_KEY_SILENT], arguments[EXARG_KEY_SMOKE], arguments) & COMSIG_CANCEL_EXPLOSION)
```
I sure hope so, but so far I haven't found one.

TemporalOroboros replied:
Changed it to use the arguments list instead, since we're already passing that. It won't look nearly as good on the hook's end but whatever.

TemporalOroboros commented on code/modules/reagents/reagent_dispenser.dm:
```diff
@@ -193,7 +193,7 @@
 
 /obj/structure/reagent_dispensers/beerkeg/blob_act(obj/structure/blob/B)
-	explosion(src.loc,0,3,5,7,10)
+	explosion(src, devastation_range = 0, heavy_impact_range = 3, light_impact_range = 5, flash_range = 7, flame_range = 10)
```
Yes.

TemporalOroboros replied:
Fixed

TemporalOroboros commented on code/modules/projectiles/projectile/special/rocket.dm:
```diff
@@ -65,9 +65,9 @@
 	if(!isliving(target)) //if the target isn't alive, so is a wall or something
-		explosion(target, 0, 1, 2, 4, flame_range = 3)
+		explosion(target, devastation_range = 0, heavy_impact_range = 1, light_impact_range = 2, flash_range = 4, flame_range = 3)
 	else
-		explosion(target, 0, 0, 2, 4, flame_range = 3)
+		explosion(target, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 2, flash_range = 4, flame_range = 3)
```
Fixed

TemporalOroboros commented on code/modules/projectiles/projectile/special/rocket.dm:
```diff
@@ -47,9 +47,9 @@
 	if(!isliving(target)) //if the target isn't alive, so is a wall or something
-		explosion(target, 0, 1, 2, 4, flame_range = 3)
+		explosion(target, devastation_range = 0, heavy_impact_range = 1, light_impact_range = 2, flash_range = 4, flame_range = 3)
 	else
-		explosion(target, 0, 0, 2, 4, flame_range = 3)
+		explosion(target, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 2, flash_range = 4, flame_range = 3)
```
Fixed

TemporalOroboros commented on code/modules/projectiles/projectile/special/rocket.dm:
```diff
@@ -24,7 +24,7 @@
 /obj/projectile/bullet/a84mm/on_hit(atom/target, blocked = FALSE)
 	..()
-	explosion(target, -1, 1, 3, 1, 0, flame_range = 4)
+	explosion(target, devastation_range = -1, heavy_impact_range = 1, light_impact_range = 3, flash_range = 1, flame_range = 4, adminlog = FALSE)
```
Fixed

TemporalOroboros commented on code/modules/projectiles/pins.dm:
```diff
@@ -69,7 +69,7 @@
 			user.show_message("<span class='danger'>SELF-DESTRUCTING...</span><br>", MSG_VISUAL)
 			to_chat(user, "<span class='userdanger'>[gun] explodes!</span>")
-		explosion(get_turf(gun), -1, 0, 2, 3)
+		explosion(src, devastation_range = -1, heavy_impact_range = 0, light_impact_range = 4, flash_range = 3)
```
I don't know. It's fixed.

TemporalOroboros commented on code/modules/projectiles/guns/magic/wand.dm:
```diff
@@ -241,7 +241,7 @@
 /obj/item/gun/magic/wand/fireball/zap_self(mob/living/user)
 	..()
-	explosion(user.loc, -1, 0, 2, 3, 0, flame_range = 2)
+	explosion(user, devastation_range = -1, heavy_impact_range = 0, light_impact_range = 2, flame_range = 2, flash_range = 3, adminlog = 0)
```
Fixed

TemporalOroboros commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/goliath_broodmother.dm:
```diff
@@ -199,7 +199,7 @@
 		mother.children_list -= src
 	visible_message("<span class='warning'>[src] explodes!</span>")
-	explosion(get_turf(loc),0,0,0,flame_range = 3, adminlog = FALSE)
+	explosion(src, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 0, flame_range = 3, adminlog = FALSE)
```
Not really, I'll get rid of them.

TemporalOroboros replied:
Done.

TemporalOroboros commented on code/game/gamemodes/meteor/meteors.dm:
```diff
@@ -249,7 +249,7 @@ GLOBAL_LIST_INIT(meteorsC, list(/obj/effect/meteor/dust)) //for space dust event
 /obj/effect/meteor/big/meteor_effect()
 	..()
-	explosion(src.loc, 1, 2, 3, 4, 0)
+	explosion(src, devastation_range = 1, heavy_impact_range = 2, light_impact_range = 3, flame_range = 0, flash_range = 4, adminlog = FALSE)
```
Done

TemporalOroboros committed:
Handles most of @LemonInTheDark's requests.

- Removes a whole bunch of argname = 0 in explosion calls.
- Un-reverses the order of flame and flash range in a few places.

TemporalOroboros committed:
Removes named callback arguments.

TemporalOroboros committed:
Changes the explosion signals to just use the arguments list

TemporalOroboros committed:
Fix signal hook arguments.

TemporalOroboros commented on code/game/objects/structures/training_machine.dm:
```diff
@@ -49,7 +49,7 @@
 /obj/structure/training_machine/obj_destruction(damage_flag)
 	remove_attached_item(throwing = TRUE)
-	explosion(src, 0,0,1, flame_range = 2)
+	explosion(src, devastation_range = 0, heavy_impact_range = 0, light_impact_range = 1, flash_range = 2)
```
Fixed

LemonInTheDark commented on code/modules/research/anomaly/explosive_compressor.dm:
```diff
@@ -80,65 +104,117 @@
+		end_test("ERROR: Missing equpment. Items ejected.")
+		return
+
+	say("Beginning compression test. Opening transfer valve.")
```
will this pop up as maptext?

TemporalOroboros replied:
Yes.

LemonInTheDark commented on code/modules/research/anomaly/explosive_compressor.dm:
```diff
@@ -80,65 +104,117 @@
+		inserted_bomb = null
+	if(timeout_timer)
+		QDEL_NULL(timeout_timer)
+	if(message)
+		say(message)
```
Ditto here

TemporalOroboros replied:
Yep.

LemonInTheDark commented:
Fix the merge conflicts and we're golder

LemonInTheDark commented:
Oh also fix #58717

TemporalOroboros committed:
Merge branch 'master' into internal-explosions

TemporalOroboros committed:
Fixes explosions not exploding
